lib/internal/encoding.js

```diff
@@ -0,0 +1,461 @@
+      throw new errors.RangeError('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    var flags = 0;
+    if (options !== null) {
+      flags |= options.fatal ? CONVERTER_FLAGS_FATAL : 0;
+      flags |= options.ignoreBOM ? CONVERTER_FLAGS_IGNORE_BOM : 0;
+    }
+
+    const handle = getConverter(enc, flags);
+    if (handle === undefined)
+      throw new errors.Error('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    this[kHandle] = handle;
+    this[kFlags] = flags;
+    this[kEncoding] = enc;
+  }
+
+  get encoding() {
+    if (this == null || this[kDecoder] !== true)
+      throw new errors.TypeError('ERR_INVALID_THIS', 'TextDecoder');
+    return this[kEncoding];
+  }
+
+  get fatal() {
+    if (this == null || this[kDecoder] !== true)
+      throw new errors.TypeError('ERR_INVALID_THIS', 'TextDecoder');
+    return (this[kFlags] & CONVERTER_FLAGS_FATAL) === CONVERTER_FLAGS_FATAL;
+  }
+
+  get ignoreBOM() {
+    if (this == null || this[kDecoder] !== true)
+      throw new errors.TypeError('ERR_INVALID_THIS', 'TextDecoder');
+    return (this[kFlags] & CONVERTER_FLAGS_IGNORE_BOM) ===
+           CONVERTER_FLAGS_IGNORE_BOM;
+  }
+
+  decode(input = empty, options = {}) {
+    if (this == null || this[kDecoder] !== true)
+      throw new errors.TypeError('ERR_INVALID_THIS', 'TextDecoder');
+    if (isArrayBuffer(input)) {
+      input = lazyBuffer().from(input);
+    } else if (!ArrayBuffer.isView(input)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'input',
+                                 ['ArrayBuffer', 'ArrayBufferView']);
+    }
+    if (typeof options !== 'object') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'options', 'object');
+    }
+
+    var flags = 0;
+    if (options !== null)
+      flags |= options.stream ? 0 : CONVERTER_FLAGS_FLUSH;
+
+    const ret = _decode(this[kHandle], input, flags);
+    if (typeof ret === 'number') {
+      const err = new errors.TypeError('ERR_ENCODING_INVALID_ENCODED_DATA',
+                                       this.encoding);
+      err.errno = ret;
+      throw err;
+    }
+    return ret.toString('ucs2');
+  }
+
+  [inspect](depth, opts) {
+    if (this == null || this[kDecoder] !== true)
+      throw new errors.TypeError('ERR_INVALID_THIS', 'TextDecoder');
+    if (typeof depth === 'number' && depth < 0)
+      return opts.stylize('[Object]', 'special');
+    var ctor = getConstructorOf(this);
+    var obj = Object.create({
+      constructor: ctor === null ? TextDecoder : ctor
+    });
+    obj.encoding = this.encoding;
+    obj.fatal = this.fatal;
+    obj.ignoreBOM = this.ignoreBOM;
+    if (opts.showHidden) {
+      obj[kFlags] = this[kFlags];
+      obj[kHandle] = this[kHandle];
+    }
+    return util.inspect(obj, opts);
+  }
+}
+
+class TextEncoder {
+  constructor() {
+    if (!warned) {
+      warned = true;
+      process.emitWarning(experimental, 'ExperimentalWarning');
+    }
+  }
+
+  get encoding() {
+    if (this == null || this[kEncoder] !== true)
```

Comment:
This won't error out on `Object.create(TextEncoder.prototype)`... sorry, the Web IDL spec is extra verbose.